src/com/google/enterprise/adaptor/DocIdPusher.java

```diff
@@ -184,6 +184,43 @@ public GroupPrincipal pushGroupDefinitions(
       throws InterruptedException;
 
   /**
+   * Blocking call to push group definitions to GSA ends in success or
+   * when provided error handler gives up.  Can take significant time
+   * if errors arise.
+   *
+   * <p>A group definition consists of a group being defined
+   * and members, which is a list of users and groups.
+   *
+   * <p>If you plan on using the return code, then the provided map should have
+   * a predictable iteration order, like {@link java.util.TreeMap}.
+   *
+   * <p>An incremental push augments or updates any existing group definitions
+   * for the groupSource on the GSA. Otherwise, the supplied group definitions
+   * will wholly replace all existing definitions for the groupSource on the
+   * GSA.
+   *
+   * <p>A groupSource identifies the source of the group definitions. If
+   * provided, it must be a string of the form {@code [a-zA-Z_][a-zA-Z0-9_-]*}.
+   * If groupSource is {@code null}, then the {@code feed.name} configuration
+   * property is used.
+   *
+   * <p>If handler is {@code null}, then a default error handler is used.
+   *
+   * @param defs map of group definitions
+   * @param caseSensitive when comparing Principals
+   * @param incremental if {@code true} incremental update is done, otherwise
+   *        full replacement is done.
+   * @param groupSource
+   * @param handler for dealing with errors pushing
+   * @return {@code null} on success, otherwise the first GroupPrincipal to fail
+   * @throws InterruptedException if interrupted and no definitions were sent
+   */
+  public GroupPrincipal pushGroupDefinitions(
```

Comment:
The original change removed an existing method from the interface, which broke binary and source compatibility with production code. This change breaks source compatibility with unit tests, which is a much smaller change.

The suggested fix is to put the classes that would need modifying in the connectors into the library where they can be updated as needed in the future. (This was proposed and rejected long ago.) Making DocIdPusher an abstract class instead of an interface would also solve the problem in the future. Using a Java 8 default method would be source and binary compatible, though both solutions are a little odd.

The solution through reflection is a little more complicated than just instanceof and a cast, to accommodate the mocks that cause the source incompatibility in the first place. We would need to use reflection to look for and invoke the new method. Making an ugly change in the production code of most connectors seems clearly worse than making a rote change in the unit tests of most connectors.